docs: bold-updates (#1988)
OffchainLabs/arbitrum-docs: arbitrum-docs/how-arbitrum-works/07-interactive-fraud-proofs.mdx

content_type: get-started
---

This document attempts to give an account of BoLD, Arbitrum's new all-vs-all challenge protocol, in terms that might be familiar to people used to studying Arbitrum's previous, 1-vs-1 challenge protocol. For another explanation of BoLD's concepts, refer to [Gentle Introduction](/how-arbitrum-works/bold/gentle-introduction.mdx), [Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx), or the BoLD [Research whitepaper](https://arxiv.org/abs/2404.10491).

> Suppose the rollup chain looks like this:

Reviewer: ```suggestion
Suppose the Rollup chain looks like this:
```

content_type: get-started
---

This document attempts to give an account of BoLD, Arbitrum's new all-vs-all challenge protocol, in terms that might be familiar to people used to studying Arbitrum's previous, 1-vs-1 challenge protocol. For another explanation of BoLD's concepts, refer to [Gentle Introduction](/how-arbitrum-works/bold/gentle-introduction.mdx), [Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx), or the BoLD [Research whitepaper](https://arxiv.org/abs/2404.10491).

Suppose the Rollup chain looks like this:

![600px-img](../assets/interactive-fraud-proof.svg)

RBlocks 93 and 95 are siblings (they both have 92 as predecessor). Alice is bonded on 93 and Bob is bonded on 95.

At this point we know that Alice and Bob disagree about the correctness of <a data-quicklook-from="rblock">RBlock</a> 93, with Alice committed to 93 being correct and Bob committed to 93 being incorrect. (Bob is bonded on 95, and 95 implicitly claims that 92 is the last correct RBlock before it, which implies that 93 must be incorrect.)

> Whenever two (or more) bonders are bonded on sibling RBlocks, a challenge begins between all of them. (There can only be one bonder per RBlock, but multiple RBlocks may participate in a challenge.) The Rollup protocol will record and referee the challenge, eventually declaring a winner and confiscating the losers' bonds. This will cause the losers to be removed as bonders.

Author: will update

:::note

If you’re a stickler for mathematical precision, we invite you to read the [BoLD research whitepaper](https://arxiv.org/abs/2404.10491). Formalizing these and other arguments pertaining to BoLD is quite tricky and subtle!

:::

## The real bisection protocol

The real bisection protocol is conceptually similar to the simplified one described above, but with several changes that improve efficiency or deal with necessary corner cases. Here is a list of the differences.

> **Bisection over L2 blocks, then over "big steps" of many instructions, then over single instructions: **Alice's assertion is over an RBlock, which asserts the result of creating some number of Layer 2 Nitro blocks. Bisection first occurs over these Layer 2 blocks, to narrow the dispute down to a dispute about a single Layer 2 Nitro block. Next, the dispute transforms into a dispute about executing a "big step" of 2^26 instructions (this intermediate level is needed to reduce the workload on the validator). At this point, the dispute transforms into a dispute about a single execution of the <a data-quicklook-from="state-transition-function">State Transition Function</a> or in other words about the execution of a sequence of WAVM instructions. The protocol then executes the recursive bisection sub-protocol again, this time over WAVM instructions, to narrow the dispute to a single instruction. For details about how these "subchallenges" work, see [BoLD: Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx)

Author: this update is pending in another PR :)

At this point we know that Alice and Bob disagree about the correctness of <a data-quicklook-from="rblock">RBlock</a> 93, with Alice committed to 93 being correct and Bob committed to 93 being incorrect. (Bob is bonded on 95, and 95 implicitly claims that 92 is the last correct RBlock before it, which implies that 93 must be incorrect.)

> Whenever two (or more) bonders are bonded on sibling RBlocks, a challenge begins between all of them. (There can only be one bonder per RBlock, but multiple RBlocks may participate in a challenge.) The rollup protocol will record the challenge and referee it, eventually declaring a winner and confiscating the losers' bonds. This will cause the losers to be removed as bonders.

Reviewer: ```suggestion
Whenever two (or more) bonders are bonded on sibling RBlocks, a challenge begins between all of them. (There can only be one bonder per RBlock, but multiple RBlocks may participate in a challenge.) The Rollup protocol will record and referee the challenge, eventually declaring a winner and confiscating the losers' bonds. This will cause the losers to be removed as bonders.
```

We’ll describe the bisection part of the protocol twice. First, we’ll give a simplified version which is easier to understand. Then we’ll describe some ways in which the real version differs from the simplified one. For an even more thorough explanation, please see the [BoLD research whitepaper](https://arxiv.org/abs/2404.10491).

## Bisection Protocol: Simplified Version

Alice (or one of her allies - for the sake of brevity we'll refer to this party as Alice) is defending the claim that starting with the state in the predecessor RBlock, the state of the Virtual Machine can advance to the state specified in RBlock A. Once a challenge starts, Alice has to provide a commitment to the entire history of the computation: that is, to the state after each step of execution.

Alice’s first move requires her to bisect her claims about intermediate states between the beginning (0 instructions executed) and the end (N instructions executed). So we require Alice to divide her claim in half, and post the state at the half-way point, after N/2 instructions have been executed.

> Now Alice has effectively bisected her N-step <a data-quicklook-from="assertion">Assertion</a> into two (N/2)-step assertions. Bob (or one of his allies; for brevity, Bob) needs to do the same for his assertion, also deriving two new (N/2)-step assertions. At most one of these will be a "rival" of one of Alice's assertion. Two assertions are "rivals" if they agree on a prefix of the state history to which they commit but disagree on the rest; this is very similar to the concept of "siblings" for RBlocks. Whichever of the two of Alice's assertions now has a rival, Alice will proceed to defend it.

Reviewer: ```suggestion
Now Alice has effectively bisected her `N`-step <a data-quicklook-from="assertion">Assertion</a> into two (`N/2`)-step assertions. Bob (or one of his allies; for brevity, Bob) needs to do the same for his assertion, also deriving two new (`N/2`)-step assertions. At most one of these will be a "rival" of one of Alice's assertions. Two assertions are "rivals" if they agree on a prefix of the state history to which they commit but disagree on the rest; this is very similar to the concept of "siblings" for RBlocks. Whichever of the two of Alice's assertions now has a rival, Alice will proceed to defend it.
```

> At this point we’re effectively back in the original situation: Alice having made an assertion that Bob disagrees with. But we have cut the size of the assertion in half, from N to N/2. We can apply the same method again, with Alice bisecting and Bob choosing one of the halves, to reduce the size to N/4. And we can continue bisecting, so that after a logarithmic number of rounds Alice and Bob will be disagreeing about a single step of execution. That’s where the bisection phase of the protocol ends, and Alice must make a one-step proof which will be checked by the EthBridge.

Reviewer: ```suggestion
At this point we’re effectively back in the original situation: Alice having made an assertion that Bob disagrees with. But we have cut the size of the assertion in half, from `N` to `N/2`. We can apply the same method again, with Alice bisecting and Bob choosing one of the halves, to reduce the size to `N/4`. And we can continue bisecting, so that after a logarithmic number of rounds Alice and Bob will be disagreeing about a single step of execution. That’s where the bisection phase of the protocol ends, and Alice must make a one-step proof which will be checked by the EthBridge.
```

## Why Bisection Correctly Identifies a Cheater

Before talking about the complexities of the real <a data-quicklook-from="challenge-protocol">Challenge protocol</a>, let’s stop to understand why the simplified version of the protocol is correct. Here correctness means two things: (1) if Alice’s initial claim is correct, Alice (or her ally) can always win the challenge, and (2) if Alice’s initial claim is incorrect, and there is at least one honest validator participating in the protocol, Alice and her allies never win the challenge.

To prove (1), observe that if Alice’s initial claim is correct, she can offer a truthful midpoint claim, and both of the implied half-size claims will be correct. So whichever half Bob objects to, Alice will again be in the position of defending a correct claim. At each stage of the protocol, Alice will be defending a correct claim. At the end, Alice will have a correct one-step claim to prove, so that claim will be provable and Alice can win the challenge.

> To prove (2), observe that if Alice’s initial claim is incorrect, this can only be because her claimed endpoint after N steps is incorrect. Now when Alice offers her midpoint state claim, that midpoint claim is either correct or incorrect. If it’s incorrect, then Bob can force Alice to defend her first-half claim, which will be incorrect. If Alice’s midpoint state claim is correct, then her second-half claim must be incorrect, and Bob can force her to defend that. So whatever Alice does, Bob will be able to challenge an incorrect half-size claim. At each stage of the protocol, Bob can identify an incorrect claim to challenge. At the end, Alice will have an incorrect one-step claim to prove, which she will be unable to do, so Bob can win the challenge.

Reviewer: ```suggestion
To prove (2), observe that if Alice’s initial claim is incorrect, this can only be because her claimed endpoint after `N` steps is incorrect. Now when Alice offers her midpoint state claim, that midpoint claim is either correct or incorrect. If it’s incorrect, then Bob can force Alice to defend her first-half claim, which will be incorrect. If Alice’s midpoint state claim is correct, then her second-half claim must be incorrect, and Bob can force her to defend that. So whatever Alice does, Bob will be able to challenge an incorrect half-size claim. At each stage of the protocol, Bob can identify an incorrect claim to challenge. At the end, Alice will have an incorrect one-step claim to prove, which she will be unable to do, so Bob can win the challenge.
```

::: note

If you’re a stickler for mathematical precision, we invite you to read the [BoLD research whitepaper](https://arxiv.org/abs/2404.10491). Formalizing these and other arguments pertaining to BoLD is quite tricky and subtle!

:::

> ## The Real Bisection Protocol

Reviewer: ```suggestion
## The real bisection protocol
```

The real bisection protocol is conceptually similar to the simplified one described above, but with several changes that improve efficiency or deal with necessary corner cases. Here is a list of the differences.

> **Bisection over L2 blocks, then over "big steps" of many instructions, then over single instructions: **Alice's assertion is over an RBlock, which asserts the result of creating some number of Layer 2 Nitro blocks. Bisection first occurs over these Layer 2 blocks, to narrow the dispute down to a dispute about a single Layer 2 Nitro block. Next, the dispute transforms into a dispute about executing a "big step" of 2^26 instructions (this intermediate level is needed to reduce the workload on the validator). At this point, the dispute transforms into a dispute about a single execution of the <a data-quicklook-from="state-transition-function">State Transition Function</a> or in other words about the execution of a sequence of WAVM instructions. The protocol then executes the recursive bisection sub-protocol again, this time over WAVM instructions, to narrow the dispute to a single instruction. For details about how these "subchallenges" work, see [BoLD: Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx)
> **Time Limits:** The protocol can be seen as tracking a timer for each assertion, counting (essentially) how much time has passed while that assertion could be seen as the presumptive winner of the game (that is, the amount of time during which the onus is on parties opposing the assertion to act.) Think of the time allowance as being about a week. Once an assertion's timer reaches this time allowance, it can be confirmed. The time allowance is designed to give the honest parties in the protocol enough time to act, icnlduign the time needed to calculate, submit, and process moves, as well as to account for possible censorship on L1.

Reviewer: ```suggestion
**Bisection over L2 blocks, then over "big steps" of many instructions, then over single instructions: **Alice's assertion is over an RBlock, which asserts the result of creating some number of Layer 2 Nitro blocks. Bisection first occurs over these Layer 2 blocks, to narrow the dispute down to a dispute about a single Layer 2 Nitro block. Next, the dispute transforms into a dispute about executing a "big step" of 2^26 instructions (this intermediate level is needed to reduce the workload on the validator). At this point, the dispute transforms into a dispute about a single execution of the <a data-quicklook-from="state-transition-function">State Transition Function</a> or in other words about the execution of a sequence of WAVM instructions. The protocol then executes the recursive bisection sub-protocol again, this time over WAVM instructions, to narrow the dispute to a single instruction. For details about how these "subchallenges" work, see [BoLD: Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx)
**Time Limits:** The protocol can be seen as tracking a timer for each assertion, counting (essentially) how much time has passed while that assertion could be seen as the presumptive winner of the game (that is, the amount of time during which the onus is on parties opposing the assertion to act.) Think of the time allowance as being about a week. Once an assertion's timer reaches this time allowance, it can be confirmed. The time allowance is designed to give the honest parties in the protocol enough time to act, including the time needed to calculate, submit, and process moves, as well as to account for possible censorship on L1.
```

It should be clear that these changes don’t affect the basic correctness of the challenge protocol. They do, however, improve its efficiency and enable it to handle all of the cases that can come up in practice.

## Efficiency

> The challenge protocol is designed so that the dispute can be resolved with a minimum of work required by the protocol (via its Layer 1 Ethereum contracts) in its role as referee. The protocol needs to track information for the assertions created by parties in the challenge, including their history commitments and timers that can be used to calculate how long they have gone without a rival. The protocol doesn’t need to pay attention to whether those claims are correct in any way; it only needs to know whether Alice’s move “has the right shape”. For more information on what the protocol needs to keep track of, see [BoLD: Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx)

Reviewer: ```suggestion
The challenge protocol is designed to resolve the dispute with a minimum of work required by the protocol (via its Layer 1 Ethereum contracts) in its role as referee. The protocol needs to track information for the assertions created by parties in the challenge, including their history commitments and timers that can be used to calculate how long they have gone without a rival. The protocol doesn’t need to pay attention to whether those claims are correct in any way; it only needs to know whether Alice’s move “has the right shape”. For more information on what the protocol needs to keep track of, see [BoLD: Technical deep dive](/bold/concepts/bold-technical-deep-dive.mdx)
```

This operates similarly to a bisection in that the responder must provide a competing global state and machine state,
but it uses that information to transition to the execution challenge phase.

### "Big-Step" execution challenge

> Once narrowed down to an individual block, the actual machine execution can be bisected. We begin by bisecting the machine execution until we reach a "chunk" of 2^26 WASM instructions. Once the disagreement has reached this granularity, we can proceed with a "small-step" execution challenge ([see below](#small-step-execution-challenge)). For more detail on why we use two levels of execution challenges ("big step" and "small-step"), see "[Challenge Resolution](/bold/concepts/bold-technical-deep-dive.mdx#challenge-resolution)" in the BoLD technical deep dive.

Reviewer 2: ```suggestion
Once narrowed down to an individual block, the actual machine execution can be bisected. We begin by bisecting the machine execution until we reach a "chunk" of 2^26 `WASM` instructions. Once the disagreement has reached this granularity, we can proceed with a "small-step" execution challenge ([see below](#small-step-execution-challenge)). For more detail on why we use two levels of execution challenges ("big step" and "small-step"), see "[Challenge Resolution](/bold/concepts/bold-technical-deep-dive.mdx#challenge-resolution)" in the BoLD technical deep dive.
```

> ### "Big-Step" Execution challenge
> 
> Once narrowed down to an individual block, the actual machine execution can be bisected. We begin by bisecting tghe machine execution until we reach a "chunk" of 2^26 WASM instructions. Once the disagreement has reached this granularity, we can proceed with a "small-step" execution challenge ([see below](#small-step-execution-challenge)). For more detail on why we use two levels of execution challenges ("big step" and "small-step"), see "[Challenge Resolution](/bold/concepts/bold-technical-deep-dive.mdx#challenge-resolution)" in the BoLD technical deep dive.

Reviewer: ```suggestion
### "Big-Step" execution challenge

Once narrowed down to an individual block, the actual machine execution can be bisected. We begin by bisecting the machine execution until we reach a "chunk" of 2^26 WASM instructions. Once the disagreement has reached this granularity, we can proceed with a "small-step" execution challenge ([see below](#small-step-execution-challenge)). For more detail on why we use two levels of execution challenges ("big step" and "small-step"), see "[Challenge Resolution](/bold/concepts/bold-technical-deep-dive.mdx#challenge-resolution)" in the BoLD technical deep dive.
```

### "Small-Step" Execution challenge

> After reaching a "chunk" of 2^26 instructions, we begin a new challenge level, bisecting this chunk until we reach a dispute over a single instruction. Once the execution has been bisected down to an individual step, `oneStepProveExecution` can be called by any player to resolve the one-step dispute. That player must provide proof data to execute a step of the machine. This proof is checked on L1. If it is correct, the one-step edge to which the one-step proof applied is considered confirmed.

Author: ```suggestion
After reaching a "chunk" of 2^26 instructions, we begin a new challenge level, bisecting this chunk until we reach a dispute over a single instruction. Once the execution has been bisected down to an individual step, `oneStepProveExecution` can be called by any player to resolve the one-step dispute. That player must provide proof data to execute a step of the machine. This proof is checked on L1. If it is correct, the one-step edge to which the one-step proof is applied and considered confirmed.
```

> ### "Small-Step" Execution challenge
> 
> After reaching a "chunk" of 2^26 instructions, we begin a new challenge level, bisecting this chunk until we reach a dispute over a single instruction. Once the execution has been bisected down to an individual step, `oneStepProveExecution` can be called by any player to resolve the one-step dispute. That player must provide proof data to execute a step of the machine. This proof is checked on L1. If it is correct, the one-step edge to which the one-step proof applied is considered confirmed.

Reviewer: ```suggestion
### "Small-Step" Execution challenge

After reaching a "chunk" of 2^26 instructions, we begin a new challenge level, bisecting this chunk until we reach a dispute over a single instruction. Once the execution has been bisected down to an individual step, `oneStepProveExecution` can be called by any player to resolve the one-step dispute. That player must provide proof data to execute a step of the machine. This proof is checked on L1. If it is correct, the one-step edge to which the one-step proof is applied is considered confirmed.
```

Author: proposed new verbiage

If a case is "unreachable", that is, the case is assumed to never arise in correct execution,
then the OSP can implement any instruction semantics in that case.

- In a challenge between malicious assertions, that start from an unreachable machine state, any case
  can arise. The challenge protocol must do
  something safe in every case. But the instruction semantics can be weird in such cases because

>   no honest party ever needs to try to refute an assertion that starts from an unreachable state. All assertions engaging in a dispute starting from an unreachable state must be malicious, and the protocol doesn't need to care which one of them wins. In particular, it is OK for the onchain one-step proof contract to differ in its behavior from the node software's execution engine in this case (and _only_ in this case).

Reviewer: ```suggestion
  no honest party ever needs to try to refute an assertion that starts from an unreachable state. All assertions engaging in a dispute starting from an unreachable state must be malicious, and the protocol doesn't need to care which one wins. In particular, it is OK for the onchain one-step proof contract to differ in its behavior from the node software's execution engine in this case (and _only_ in this case).
```
